traversal/procedure/ProcedureVertex.java

```diff
@@ -175,124 +189,347 @@ <ORDER extends Order> Forwardable<? extends ThingVertex, ORDER> iterateAndFilter
             assert id().isVariable();
-            Set<Traversal.Parameters.Value> values = parameters.getValues(id().asVariable(), eqPredicate);
-            if (values.size() > 1) return emptySorted(order);
-            Traversal.Parameters.Value value = values.iterator().next();
-            TreeSet<AttributeVertex<?>> attributes = new TreeSet<>();
+            Set<Traversal.Parameters.Value<?>> values = parameters.getValues(id().asVariable(), eqPredicate);
+            if (values.size() > 1) return list();
+            Traversal.Parameters.Value<?> value = values.iterator().next();
+            List<AttributeVertex<?>> attributes = new ArrayList<>();
             attributeTypes.map(t -> attributeVertex(graphMgr, t, value))
                     .filter(Objects::nonNull).forEachRemaining(attributes::add);
-            return iterateSorted(attributes, order);
+            return attributes;
         }
 
-        private AttributeVertex<?> attributeVertex(GraphManager graphMgr, TypeVertex type, Traversal.Parameters.Value value) {
+        private <T> AttributeVertex<?> attributeVertex(
+                GraphManager graphMgr, TypeVertex type, Traversal.Parameters.Value<T> value
+        ) {
             assert type.isAttributeType();
             Encoding.ValueType<?> valueType = type.valueType();
-            if (valueType == BOOLEAN) return graphMgr.data().getReadable(type, value.getBoolean());
-            else if (valueType == LONG) return graphMgr.data().getReadable(type, value.getLong());
-            else if (valueType == DOUBLE) return graphMgr.data().getReadable(type, value.getDouble());
-            else if (valueType == STRING) return graphMgr.data().getReadable(type, value.getString());
-            else if (valueType == DATETIME) return graphMgr.data().getReadable(type, value.getDateTime());
+            if (valueType == BOOLEAN) {
+                return graphMgr.data().getReadable(type, convertToBoolean(value.valueType(), value.value()));
+            } else if (valueType == LONG) {
+                return graphMgr.data().getReadable(type, convertToLong(value.valueType(), value.value(), true));
+            } else if (valueType == DOUBLE) {
+                return graphMgr.data().getReadable(type, convertToDouble(value.valueType(), value.value()));
+            } else if (valueType == STRING) {
+                return graphMgr.data().getReadable(type, convertToString(value.valueType(), value.value()));
+            } else if (valueType == DATETIME) {
+                return graphMgr.data().getReadable(type, convertToDateTime(value.valueType(), value.value()));
+            }
             throw TypeDBException.of(ILLEGAL_STATE);
         }
 
-        static <ORDER extends Order> Forwardable<AttributeVertex<?>, ORDER> filterAttributes(Forwardable<? extends ThingVertex, ORDER> iterator) {
-            return mapToAttributes(iterator.filter(ThingVertex::isAttribute));
+        private <VALUE> ThingVertex attributeVertexTarget(
```

Comment:
we can now create a virtual `Target` vertex, for which we construct the expected IID for the value that we are looking for. This is used for skipping through ranges of values when executing an optimised predicate.